### PR TITLE
Marketing Connections: Fix the tour position

### LINKED ISSUE #44194: Marketing: Rogue guided tour
I was clicking through the different sub tabs of the marketing section and a guided tour appeared. I suppose I clicked quickly enough to a new tab of the marketing section and it caused this to happen:

image

That tour tip stayed on the screen (at x0,y0 presumably because the div it was supposed to target was removed from the dom when I switched tabs?) as I clicked from tab to tab. 

Larger question is do we need to use a guide tour here or could we have a tooltip appear on hover or something else on the sharing screen?

### PR DESCRIPTION
**Pull request by pablinos**
#### Changes proposed in this Pull Request

The tour displayed on the marketing connections page was being
incorrectly positioned at the top left. This was due to the target
element not existing until the connection statuses had been received for
the API

The guided tour API has a `wait` property, which is a function that will
prevent the guide from appearing until it returns. This change copies a
pattern from another Jetpack guide, which polls for the existence of the
element. I'm not convinced that's a great solution. I think it would be
better to have the guide itself wait for the existence of the target
element rather than default to the top left of the screen. This could
have other consequences, and so this fix seems reasonable for now, until
this can be looked at more closely.

The `meta` settings also made this tour appear by simply visiting the
`/marketing/connections` route on a desktop. This wasn't intended and
should have be left to be triggered by the `tour` parameter added from
the Customer Home card.

Lastly, on my test site with a premium plan, a prompt to upgrade
disappeared, but the tour didn't reposition until I scrolled or resized
the window. Setting the `keepRepositioning` prop helps with this, but
that only polled once every three seconds. I've reduced that to 700ms,
so there's less jank. Opinions on this change are welcome!

Props to @allilevine for debugging this with me.

#### Testing instructions

* Go to `/marketing/connections/?tour=marketingConnectionsTour`
* Without this change, the tour should display incorrectly at the top left of the screen
* With this change, the tour should appear once the connection statuses have loaded and as the screen changes it should adjust its position.

Fixes #44194

### CODE REVIEW

File: client/layout/guided-tours/config-elements/step.tsx [hunk 1]
```diff
@@ -120,7 +120,7 @@ export default class Step extends Component< Props, State > {
 			this.watchTarget();
 		} );
 		if ( this.props.keepRepositioning ) {
-			this.repositionInterval = setInterval( this.onScrollOrResize, 3000 );
+			this.repositionInterval = setInterval( this.onScrollOrResize, 700 );
```
Comment (pablinos) [inline client/layout/guided-tours/config-elements/step.tsx:123]: I'm not sure if it is a performance issue to reduce this timeout

File: client/layout/guided-tours/tours/marketing-connections-tour/index.js [hunk 2]
```diff
@@ -10,6 +10,16 @@ import meta from './meta';
...
+const CONNECT_BUTTON_SELECTOR = '.sharing-service.not-connected .button.is-compact';
+
+// Wait until the desired DOM element appears. Check every 125ms.
+// This function is a Redux action creator, hence the two arrows.
+const waitForConnectButton = () => async () => {
+	while ( ! document.querySelector( CONNECT_BUTTON_SELECTOR ) ) {
+		await new Promise( ( resolve ) => setTimeout( resolve, 125 ) );
+	}
+};
+
```
Comment (pablinos) [inline client/layout/guided-tours/tours/marketing-connections-tour/index.js:22]: This seems like a hacky solution, but it is [a pattern used elsewhere](https://github.com/Automattic/wp-calypso/blob/master/client/layout/guided-tours/tours/jetpack-plugin-updates-tour/index.tsx#L27) so perhaps it's ok for now here too?

**Comment by creativecoder** [comment]
+1 to avoiding timeouts in favor a MutationObserver, if possible.

**Comment by pablinos** [comment]
I've changed this so the `MutationObserver` triggers the updates. The original polling using for `keepRepositioning` didn't provide the greatest UX. The only other place it is used is the `simplePaymentsEmailTour`. This tip points to the Pages link in the sidebar. Both this implementation and the previous one suffer from a problem of not displaying if that menu is collapsed.

The tour is triggered by a link in a marketing email, which I'm trying to work if we are still sending. If we are then we should update the link to point to Customer Home rather than the stats page, and if the `tour` query param is present, dispatch an event to ensure that the Site sidebar section is open.

I'd say it's probably ready for a proper review now. If you'd like to take another look @aaronrobertshaw, that would be much appreciated!

**Comment by aaronrobertshaw** [comment]
@pablinos I've only gotten a brief chance to take a look at this one again. Looking good so far!

Previously, if we visited `/marketing/connections/?tour=marketingConnectionsTour` and then selected a site, the tour was shown. Can we be sure that the only links people will be following will included the site slug? This might be answered if we can work out if the marketing email you mentioned is still being sent.

I also happened to notice a tiny edge case on larger display sizes. When the viewport is `>= 2040px` wide, the tip is positioned off to the right of the button making the arrow the wrong side of the tip.

Screen Shot 2020-07-23 at 3 18 22 pm

**Comment by allilevine** [comment]
I added a "left" placement to position the tip correctly on larger screen sizes. @aaronrobertshaw Would you mind taking another look? 🙏 

> Previously, if we visited `/marketing/connections/?tour=marketingConnectionsTour` and then selected a site, the tour was shown. Can we be sure that the only links people will be following will included the site slug? This might be answered if we can work out if the marketing email you mentioned is still being sent.

@pablinos Were we able to confirm this?

**Comment by pablinos** [comment]
Thanks for the review @aaronrobertshaw!

> I'd say if this is still in use we can tweak it in separate PR

Yeah I'm inclined to agree. If it gets sent out again then we should probably be testing the flow anyway.

**Pull request merged**